### PR TITLE
Feature/logging tool

### PR DESCRIPTION
**Pull request by simonmitchell**

### CODE REVIEW

File: Baymax/Logger/Log.swift [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    case fault
+}
+
+extension DateFormatter {
+    
+    static var logFileNameFormatter: DateFormatter {
+        let formatter = DateFormatter()
+        formatter.calendar = Calendar(identifier: .iso8601)
+        formatter.locale = Locale(identifier: "en_US_POSIX")
+        formatter.dateFormat = "YYYY-MM-DD'T'hh:mm:ssZ"
```
Comment (BenShutt) [inline Baymax/Logger/Log.swift]: Might want `HH` here
Not sure on what `DD` does vs `dd`

File: Baymax/Logger/Log.swift [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        guard let url = fm.urls(for: searchPathDirectory, in: domainMask).first else { return nil }
+        return url.appendingPathComponent("baymax_logs", isDirectory: true)
+    }
+    
+    /// Initialises a new logger object with the given file name in the search path and domain mask provided
+    /// - Parameters:
+    ///   - fileName: The name to give to the log file
+    ///   - searchPathDirectory: The search path directory (defaults to documents directory)
+    ///   - domainMask: The domain mask (defaults to user domain mask)
+    public init(fileName: String, searchPathDirectory: FileManager.SearchPathDirectory = .documentDirectory, domainMask: FileManager.SearchPathDomainMask = .userDomainMask) {
```
Comment (BenShutt) [inline Baymax/Logger/Log.swift]: While might not be so important here I encourage short-width lines, so breaking up the parameters.
This is a bit easier to read from a laptop :)

File: Baymax/Logger/Log.swift [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    /// The search path directory the log will save to
+    private let searchPathDirectory: FileManager.SearchPathDirectory
+    
+    /// The domain mask the log will save to
+    private let domainMask: FileManager.SearchPathDomainMask
+    
+    private let fm = FileManager.default
+    
+    internal var directory: URL? {        
+        guard let url = fm.urls(for: searchPathDirectory, in: domainMask).first else { return nil }
```
Comment (BenShutt) [inline Baymax/Logger/Log.swift]: Think there's an additional parameter to to create the directory here. Would be good to make sure that's explicitly `true`. 
Some of the `searchPathDirectory`s are not created on first install

EDIT: I see you do the create directory below, up to you on preferred method.

File: Baymax/Logger/Log.swift (hunk 2 above)
Comment (BenShutt) [inline Baymax/Logger/Log.swift]: Feel like `documentDirectory` is for putting stuff the user should see? Often I opt for `applicationSupportDirectory` but you might be right with documents depending what you want here

File: Baymax/Logger/Log.swift [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    
+    public func write(_ string: String) {
+        
+        guard UserDefaults.standard.baymaxLoggingEnabled else {
+            return
+        }
+                
+        var writeString = string
+        
+        if !writeString.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
```
Comment (BenShutt) [inline Baymax/Logger/Log.swift]: Maybe we should exit is the `string` is whitespace

**Review by RodrigoLondon — APPROVED** [review]

File: Baymax/Logger/Log.swift (hunk 1 above)
Comment (simonmitchell) [reply Baymax/Logger/Log.swift]: You were correct, sorted now! :D

File: Baymax/Logger/Log.swift (hunk 2 above)
Comment (simonmitchell) [reply Baymax/Logger/Log.swift]: So I left this as it was with GDPC First Aid, idea there was that it made it easier for client/users to share, but you may well be right! Will discuss with PMs

**Pull request merged**